### PR TITLE
Pix2Struct: fix wrong broadcast axis of attention mask in visual encoder

### LINKED ISSUE #23974: Pix2Struct: wrong attention mask broadcasting in the visual encoder
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.10
- Python version: 3.8.16
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.6.10 (cpu)
- Jax version: 0.4.10
- JaxLib version: 0.4.10
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker @younesbelkada @amyeroberts

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The Pix2Struct visual attention layer masks the attention scores (logits) by adding negative infinities from the attention mask.
https://github.com/huggingface/transformers/blob/539e2281cd97c35ef4122757f26c88f44115fa94/src/transformers/models/pix2struct/modeling_pix2struct.py#L200-L220

And the below line is to broadcast the boolean attention mask of which shape is `[batch_size, seq_len]` to make a shape of `[batch_size, num_heads, query_len, key_len]`. But it seems the mask tensor is broadcasted on wrong axes.
https://github.com/huggingface/transformers/blob/539e2281cd97c35ef4122757f26c88f44115fa94/src/transformers/models/pix2struct/modeling_pix2struct.py#L213

Here is the common way to extend the mask:

https://github.com/huggingface/transformers/blob/539e2281cd97c35ef4122757f26c88f44115fa94/src/transformers/modeling_utils.py#L896

Because of that, the model will attend mask tokens as well while the original implementation uses `flaxformer` as follows:
https://github.com/google/flaxformer/blob/9adaa4467cf17703949b9f537c3566b99de1b416/flaxformer/components/attention/dense_attention.py#L2019

### Expected behavior

Fix the order of new axes on the attention mask.

### PR DESCRIPTION
**Pull request by affjljoo3581**
# What does this PR do?

Fixes #23974

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @younesbelkada @amyeroberts

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for flagging the issue, in principle this looks good to me
This leads to getting slightly different results at inference with the same input, thus some slow tests we had currently fails with this modification.
Is it ok if I directly push on your branch to change the expected values of the slow tests and doctests?

**Comment by affjljoo3581** [comment]
Sure. Is there something else I have to do for that?

**Comment by younesbelkada** [comment]
Should be all good! just pushed!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!
